### PR TITLE
NIFI-5400 - Changed the hostname verifier from the custom NiFi verifi…

### PR DESCRIPTION
**Pull request by thenatog**
…er to the Apache http-client DefaultHostnameVerifier

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [X] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [X] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [X] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by alopresto** [comment]
Reviewing...

**Comment by alopresto** [comment]
This PR removes the locally-defined inner class `NiFiHostnameVerifier` from `NiFiClientFactory` and adds unit tests but does not cover the class [`org.apache.nifi.web.util.NiFiHostnameVerifier`](https://github.com/apache/nifi/blob/master/nifi-commons/nifi-web-utils/src/main/java/org/apache/nifi/web/util/NiFiHostnameVerifier.java#L18) which is used in [`org.apache.nifi.web.util.WebUtils#createClientHelper()`](https://github.com/apache/nifi/blob/master/nifi-commons/nifi-web-utils/src/main/java/org/apache/nifi/web/util/WebUtils.java#L98). Please remove that class entirely and instantiate a `org.apache.http.conn.ssl.DefaultHostnameVerifier` in `WebUtils#createClientHelper()` and verify that the calling references (uses of `WebUtils#createClient(ClientConfig)` and `WebUtils#createClient(ClientConfig, SSLContext)` listed below) continue to work properly (i.e. add unit tests & regression tests for edge case hostnames). 

```
Method
    createClient(ClientConfig, SSLContext)
Found usages  (5 usages found)
    Unclassified usage  (5 usages found)
        nifi-confluent-schema-registry-service  (1 usage found)
            org.apache.nifi.confluent.schemaregistry.client  (1 usage found)
                RestSchemaRegistryClient  (1 usage found)
                    RestSchemaRegistryClient(List<String>, int, SSLContext, ComponentLog)  (1 usage found)
                        77 client = WebUtils.createClient(clientConfig, sslContext);
        nifi-framework-core  (1 usage found)
            org.apache.nifi.remote  (1 usage found)
                RemoteNiFiUtils  (1 usage found)
                    getClient(SSLContext)  (1 usage found)
                        50 client = WebUtils.createClient(clientConfig, sslContext);
        nifi-web-api  (2 usages found)
            org.apache.nifi.integration.accesscontrol  (1 usage found)
                ITAccessTokenEndpoint  (1 usage found)
                    setup()  (1 usage found)
                        88 final Client client = WebUtils.createClient(null, createTrustContext(props));
            org.apache.nifi.integration.util  (1 usage found)
                NiFiTestServer  (1 usage found)
                    getClient()  (1 usage found)
                        170 return WebUtils.createClient(null, SslContextFactory.createSslContext(properties));
        nifi-web-security  (1 usage found)
            org.apache.nifi.web.security.x509.ocsp  (1 usage found)
                OcspCertificateValidator  (1 usage found)
                    OcspCertificateValidator(NiFiProperties)  (1 usage found)
                        110 client = WebUtils.createClient(clientConfig, SslContextFactory.createSslContext(properties));
```

```
Method
    createClient(ClientConfig)
Found usages  (3 usages found)
    Unclassified usage  (3 usages found)
        nifi-framework-core  (1 usage found)
            org.apache.nifi.remote  (1 usage found)
                RemoteNiFiUtils  (1 usage found)
                    getClient(SSLContext)  (1 usage found)
                        48 client = WebUtils.createClient(clientConfig);
        nifi-web-security  (1 usage found)
            org.apache.nifi.web.security.x509.ocsp  (1 usage found)
                OcspCertificateValidator  (1 usage found)
                    OcspCertificateValidator(NiFiProperties)  (1 usage found)
                        112 client = WebUtils.createClient(clientConfig);
        nifi-yandex-processors  (1 usage found)
            org.apache.nifi.processors.yandex  (1 usage found)
                YandexTranslate  (1 usage found)
                    onScheduled(ProcessContext)  (1 usage found)
                        201 client = WebUtils.createClient(null);
```

File: nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy [hunk 1]
```diff
@@ -17,38 +17,19 @@
...
 import javax.ws.rs.client.Client
 import javax.ws.rs.client.ClientBuilder
-import java.security.KeyManagementException
-import java.security.KeyStore
-import java.security.KeyStoreException
-import java.security.NoSuchAlgorithmException
-import java.security.SecureRandom
-import java.security.UnrecoverableKeyException
-import java.security.cert.Certificate
+import java.security.*
```
Comment (alopresto) [inline nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy]: We can't import wildcards. This might be your IDE doing it automatically (in which case, loading the `checkstyle.xml` will prevent this in the future).

File: nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy [hunk 2]
```diff
@@ -17,38 +17,19 @@
...
 import java.security.cert.CertificateException
-import java.security.cert.CertificateParsingException
-import java.security.cert.X509Certificate
 
 class NiFiClientFactory implements ClientFactory{
 
-    private static final Logger logger = LoggerFactory.getLogger(NiFiClientFactory.class)
```
Comment (alopresto) [inline nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy]: Why get rid of the logger?

File: nifi-toolkit/nifi-toolkit-admin/src/test/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactorySpec.groovy [hunk 3]
```diff
@@ -121,21 +122,86 @@ class NiFiClientFactorySpec extends Specification {
...
+
+        given:
+        final String EXPECTED_DN = "CN=*.nifi.apache.org,OU=Security,O=Apache,ST=CA,C=US"
+        Certificate[] certificateChain = generateCertificateChain(EXPECTED_DN,ISSUER_DN)
+        def mockSession = Mock(SSLSession)
+        DefaultHostnameVerifier verifier = new DefaultHostnameVerifier()
+        mockSession.getPeerCertificates() >> certificateChain
+
+        when:
+        def verified = verifier.verify("client.nifi.apache.org",mockSession)
```
Comment (alopresto) [inline nifi-toolkit/nifi-toolkit-admin/src/test/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactorySpec.groovy]: Please rename these variables (and in following tests) to be more descriptive, i.e. `def validSubdomainIsVerified`, `def validSubdomainIsVerified2`, `def invalidSubdomainIsNotVerified`.

File: nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy (hunk 2 above)
Comment (thenatog) [reply nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy]: It was unused in this class after removing the NiFiHostnameVerifier

File: nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy (hunk 1 above)
Comment (thenatog) [reply nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy]: Woopsie

File: nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy (hunk 2 above)
Comment (alopresto) [reply nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy]: Ok, figured that; it's good practice to have it there anyway so that it's consistent across classes and the use stays aligned with the pattern we've established.

**Comment by alopresto** [comment]
I merged this but made two changes. First, the `nifi-web-utils` tests were not running, because they are Groovy tests and there is nothing in `src/test/java`. Without a file (even empty) in that directory, the Groovy tests do not get picked up (neither compiled nor run). I added the `groovy-eclipse-compiler` plugin to `nifi-web-utils/pom.xml` to ensure this is run. That commit is [5c0232c](https://github.com/alopresto/nifi/commit/5c0232c9dd8009dc69bc5adb1fb1ef7942832911). 

Second, there was a warning about a duplicate definition of `httpclient` dependency in `nifi-web-utils/pom.xml`. I removed it, and that commit is [5f538c6](https://github.com/alopresto/nifi/commit/5f538c69f1aebc0b6b0d6dbabf0f09c8e9854a57). 

Both of those commits were rebased onto Nathan's rebased commits as well. 

A gist demonstrating the issue is [here](https://gist.github.com/alopresto/184f3631ec44a4c036d323d622ea97aa). 

Ran `contrib-check` and all tests pass. +1, merging.
